doc/man/man3/CharacterBullet.3
.TH "CharacterBullet" 3 "Version 1.0" "Zaxxon" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CharacterBullet \- The bullet used by the player and enemies\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <CharacterBullet\&.h>\fP
.PP
Inherits \fBAbstractBullet\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBBulletDeathType\fP { \fBWallDeath\fP, \fBEnemyDeath\fP, \fBMissileDeath\fP }"
.br
.in -1c

Public Types inherited from \fBAbstractBullet\fP
.in +1c
.ti -1c
.RI "enum \fBBulletType\fP { \fBPlayer\fP, \fBEnemy\fP, \fBMissile\fP, \fBzBullet\fP, \fBxBulletL\fP, \fBxBulletR\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCharacterBullet\fP (sf::Texture *spritesheet, sf::Vector3f spawnPos, unsigned int sizeindex, \fBBulletType\fP=\fBPlayer\fP, sf::Vector2f=sf::Vector2f(0, 0))"
.br
.RI "Create a bullet based on if player or enemy shot it\&. Players use 3f & enemies use 2f, enemies run on 2f space while players are on 3f space and it's not possible to convert from 2f to 3f, so enemies have to pass their 2f position to the bullets\&. Just make sure to pass the 2f if an enemy and the 3f will be ignored\&. "
.ti -1c
.RI "void \fBkill\fP ()"
.br
.RI "Default kill, run the kill method with enemyDeath as parameter\&. "
.ti -1c
.RI "void \fBkill\fP (\fBBulletDeathType\fP type)"
.br
.RI "The bullet hit something, run code to kill bullet based on what it hit\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &window, float) override"
.br
.RI "Run the logic for this bullet and then draw it to screen\&. "
.ti -1c
.RI "unsigned int \fBgetSizeIndex\fP ()"
.br
.RI "Get public size index of this character bullet\&. "
.in -1c

Public Member Functions inherited from \fBAbstractBullet\fP
.in +1c
.ti -1c
.RI "\fBAbstractBullet\fP ()"
.br
.ti -1c
.RI "bool \fBisHit\fP ()"
.br
.RI "Returns whether the bullet is dead or not\&. "
.ti -1c
.RI "void \fBtranslate\fP (float)"
.br
.RI "Move a certain amount on the Z axis\&. "
.in -1c

Public Member Functions inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "\fBEntity\fP ()"
.br
.RI "Setup \fBEntity\fP class variables\&. "
.ti -1c
.RI "virtual \fB~Entity\fP ()=0"
.br
.RI "Clean up memory related to \fBEntity\fP class\&. "
.ti -1c
.RI "virtual void \fBkill\fP (\fBAnimation::Anim\fP \fBanimation\fP=\fBAnimation::CHARACTER_DEATH\fP)"
.br
.RI "Run death animation code for the killed entity\&. "
.ti -1c
.RI "int \fBgetAnimationState\fP ()"
.br
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f pos)"
.br
.ti -1c
.RI "sf::Vector3f \fBgetPos\fP ()"
.br
.ti -1c
.RI "sf::FloatRect \fBgetBounds\fP ()"
.br
.ti -1c
.RI "sf::Vector2f \fBgetSpritePos\fP ()"
.br
.RI "Get the position of the related sprite for this entity\&. "
.in -1c
.SS "Additional Inherited Members"


Protected Attributes inherited from \fBAbstractBullet\fP
.in +1c
.ti -1c
.RI "bool \fBalive\fP = true"
.br
.in -1c

Protected Attributes inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "sf::Texture * \fBspriteSheet\fP"
.br
.ti -1c
.RI "sf::Sprite * \fBsprite\fP"
.br
.ti -1c
.RI "\fBAnimation\fP \fBanimations\fP"
.br
.ti -1c
.RI "bool \fBdead\fP = false"
.br
.ti -1c
.RI "sf::Text \fBdebugText\fP"
.br
.ti -1c
.RI "sf::Font \fBdebugFont\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
The bullet used by the player and enemies\&. 
.PP
Definition at line \fB9\fP of file \fBCharacterBullet\&.h\fP\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBCharacterBullet::BulletDeathType\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIWallDeath \fP\fP
.TP
\fB\fIEnemyDeath \fP\fP
.TP
\fB\fIMissileDeath \fP\fP
.PP
Definition at line \fB12\fP of file \fBCharacterBullet\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "CharacterBullet::CharacterBullet (sf::Texture * spritesheet, sf::Vector3f spawnPos, unsigned int sizeIndex, \fBBulletType\fP type = \fR\fBPlayer\fP\fP, sf::Vector2f spawnPos2f = \fRsf::Vector2f(0, 0)\fP)"

.PP
Create a bullet based on if player or enemy shot it\&. Players use 3f & enemies use 2f, enemies run on 2f space while players are on 3f space and it's not possible to convert from 2f to 3f, so enemies have to pass their 2f position to the bullets\&. Just make sure to pass the 2f if an enemy and the 3f will be ignored\&. 
.PP
\fBParameters\fP
.RS 4
\fIspritesheet\fP 
.br
\fIspawnPos\fP Ignored if spawnPos2f is defined\&.
.br
\fIsizeIndex\fP Size of bullet\&.
.br
\fItype\fP 
.br
\fIspawnPos2f\fP Optional, pass through when enemy\&.
.RE
.PP

.PP
Definition at line \fB16\fP of file \fBCharacterBullet\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "unsigned int CharacterBullet::getSizeIndex ()"

.PP
Get public size index of this character bullet\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.PP
Definition at line \fB49\fP of file \fBCharacterBullet\&.cpp\fP\&.
.SS "void CharacterBullet::kill ()"

.PP
Default kill, run the kill method with enemyDeath as parameter\&. 
.PP
Definition at line \fB58\fP of file \fBCharacterBullet\&.cpp\fP\&.
.SS "void CharacterBullet::kill (\fBBulletDeathType\fP deathType)"

.PP
The bullet hit something, run code to kill bullet based on what it hit\&. 
.PP
\fBParameters\fP
.RS 4
\fIdeathType\fP 
.RE
.PP

.PP
Definition at line \fB68\fP of file \fBCharacterBullet\&.cpp\fP\&.
.SS "void CharacterBullet::update (sf::RenderWindow & window, float gameSpeed)\fR [override]\fP, \fR [virtual]\fP"

.PP
Run the logic for this bullet and then draw it to screen\&. 
.PP
\fBParameters\fP
.RS 4
\fIwindow\fP 
.RE
.PP

.PP
Implements \fBEntity\fP\&.
.PP
Definition at line \fB93\fP of file \fBCharacterBullet\&.cpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Zaxxon from the source code\&.


src/Entity/AbstractBullet/CharacterBullet/CharacterBullet.cpp
#include "CharacterBullet.h"


/// <summary>
/// Create a bullet based on if player or enemy shot it. Players use 3f &
/// enemies use 2f, enemies run on 2f space while players are on 3f space
/// and it's not possible to convert from 2f to 3f,
/// so enemies have to pass their 2f position to the bullets. Just make sure
/// to pass the 2f if an enemy and the 3f will be ignored.
/// </summary>
/// <param name="spritesheet"></param>
/// <param name="spawnPos">Ignored if spawnPos2f is defined.</param>
/// <param name="sizeIndex">Size of bullet.</param>
/// <param name="type"></param>
/// <param name="spawnPos2f">Optional, pass through when enemy.</param>
CharacterBullet::CharacterBullet(sf::Texture* spritesheet,
                                 sf::Vector3f spawnPos,
                                 unsigned int sizeIndex, BulletType type,
                                 sf::Vector2f spawnPos2f
)
{
	sprite->setTexture(*spritesheet);
	sprite->setTextureRect(sf::IntRect(8 + 16 * sizeIndex, 47, 16, 8));
	sprite->setOrigin(0, 8);

	if (type == Player)
	{
		sprite->setColor(sf::Color(0, 222, 0));
		setPos(sf::Vector3f(spawnPos.x - 21.f, spawnPos.y + .02f * sizeIndex,
		                    spawnPos.z - 15.f));
		sprite->setPosition(translateTo2d(getPos()));
	}
	else if (type == Enemy)
	{
		sprite->setColor(sf::Color(222, 0, 0));
		sprite->setPosition(
			spawnPos2f + sf::Vector2f(-3.f * static_cast<int>(3U - sizeIndex),
			                          20));
	}
	this->sizeIndex = sizeIndex;
	this->type = type;
}


/// <summary>
/// Get public size index of this character bullet
/// </summary>
/// <returns></returns>
unsigned int CharacterBullet::getSizeIndex()
{
	return sizeIndex;
}


/// <summary>
/// Default kill, run the kill method with enemyDeath as parameter
/// </summary>
void CharacterBullet::kill()
{
	kill(EnemyDeath);
}


/// <summary>
/// The bullet hit something, run code to kill bullet based on what it hit.
/// </summary>
/// <param name="deathType"></param>
void CharacterBullet::kill(BulletDeathType deathType)
{
	sprite->setColor(sf::Color(255, 255, 255));
	//death conditions and animations based on what is dying
	switch (deathType)
	{
	case WallDeath:
		if (animations.getState() < 1)
			animations.run(sprite, Animation::Anim::WALLBULLET_DEATH,
			               sizeIndex);
		break;
	case EnemyDeath:
		// adjust position to align better
		if (this->type == Player)
			setPos(getPos() + sf::Vector3f(0, 0, -16));
		if (animations.getState() < 2)
			animations.run(sprite, Animation::Anim::BULLET_DEATH, sizeIndex);
	}
}


/// <summary>
/// Run the logic for this bullet and then draw it to screen
/// </summary>
/// <param name="window"></param>
void CharacterBullet::update(sf::RenderWindow& window, float gameSpeed)
{
	if (type == Player)
	{
		if (animations.getState() == 0)
		{
			translate(-6 * gameSpeed);
		}
	}
	//doesn't use translate because it uses 2d movement
	else if (type == Enemy)
		sprite->move(translateTo2d(sf::Vector3f(0, 0, 3 * gameSpeed)));

	if (animations.getState() == 1)
		alive = false;

	window.draw(*sprite);
}
